server/src/main/java/org/opensearch/index/store/remote/filecache/FileCacheCleaner.java

```diff
@@ -52,19 +55,55 @@ public FileCacheCleaner(NodeEnvironment nodeEnvironment, FileCache fileCache) {
     @Override
     public void beforeIndexShardDeleted(ShardId shardId, Settings settings) {
         try {
-            String storeType = settings.get(IndexModule.INDEX_STORE_TYPE_SETTING.getKey());
-            if (IndexModule.Type.REMOTE_SNAPSHOT.match(storeType)) {
-                ShardPath shardPath = ShardPath.loadFileCachePath(nodeEnvironment, shardId);
-                Path localStorePath = shardPath.getDataPath().resolve(LOCAL_STORE_LOCATION);
+            if (isRemoteSnapshot(settings)) {
+                final ShardPath shardPath = ShardPath.loadFileCachePath(nodeEnvironment, shardId);
+                final Path localStorePath = shardPath.getDataPath().resolve(LOCAL_STORE_LOCATION);
                 try (DirectoryStream<Path> ds = Files.newDirectoryStream(localStorePath)) {
                     for (Path subPath : ds) {
                         fileCache.remove(subPath.toRealPath());
                     }
                 }
-                FileSystemUtils.deleteSubDirectories(shardPath.getRootDataPath());
             }
         } catch (IOException ioe) {
-            log.error(() -> new ParameterizedMessage("Error removing items from cache during shard deletion {})", shardId), ioe);
+            log.error(() -> new ParameterizedMessage("Error removing items from cache during shard deletion {}", shardId), ioe);
         }
     }
+
+    @Override
+    public void afterIndexShardDeleted(ShardId shardId, Settings settings) {
+        if (isRemoteSnapshot(settings)) {
+            final Path path = ShardPath.loadFileCachePath(nodeEnvironment, shardId).getDataPath();
+            try {
+                if (Files.exists(path)) {
+                    IOUtils.rm(path);
+                }
+            } catch (IOException e) {
+                log.error(() -> new ParameterizedMessage("Failed to delete cache path for shard {}", shardId), e);
+            }
+        }
+    }
+
+    @Override
+    public void afterIndexRemoved(
+        Index index,
+        IndexSettings indexSettings,
+        IndicesClusterStateService.AllocatedIndices.IndexRemovalReason reason
+    ) {
+        if (isRemoteSnapshot(indexSettings.getSettings())) {
```

Comment:
Should we also check for the reason `IndexRemovalReason` here? The same event is generated in case of an index being closed or any other failure.